lib/r10k/forge/module_release.rb

```diff
@@ -104,54 +111,79 @@ def download
+      # @raise [R10K::Error] when no SHA256 is available and FIPS mode is on
       # @return [void]
       def verify
         logger.debug1 "Verifying that #{@tarball_cache_path} matches checksum"
 
-        md5_of_tarball = Digest::MD5.hexdigest(File.read(@tarball_cache_path, mode: 'rb'))
+        sha256_of_tarball = Digest::SHA256.file(@tarball_cache_path).hexdigest
 
-        if @md5_file_path.exist?
-          verify_from_md5_file(md5_of_tarball)
+        if @sha256_file_path.exist?
+          verify_from_file(sha256_of_tarball, @sha256_file_path)
         else
-          verify_from_forge(md5_of_tarball)
+          forge_256_checksum = @forge_release.file_sha256
+          if forge_256_checksum
+            verify_from_forge(sha256_of_tarball, forge_256_checksum, @sha256_file_path)
+          else
+            if R10K::Util::Platform.fips?
+              raise R10K::Error, "Could not verify module, no SHA256 checksum available, and MD5 checksums not allowed in FIPS mode"
+            end
+
+            logger.debug1 "No SHA256 checksum available, falling back to MD5"
+            md5_of_tarball = Digest::MD5.file(@tarball_cache_path).hexdigest
+            if @md5_file_path.exist?
+              verify_from_file(md5_of_tarball, @md5_file_path)
+            else
+              verify_from_forge(md5_of_tarball, @forge_release.file_md5, @md5_file_path)
+            end
+          end
         end
       end
 
-      # Verify the md5 of the cached tarball against the
+      # Verify the checksum of the cached tarball against the
       # module release checksum stored in the cache as well.
       # On mismatch, remove the cached copy of both files.
+      # @param tarball_checksum [String] the checksum (either md5 or SHA256)
+      #                         of the downloaded module tarball
+      # @param file [Pathname] the file containing the checksum as downloaded
+      #             previously from the forge
+      # @param digest_class [Digest::SHA256, Digest::MD5] which checksum type
+      #                     to verify with
       #
       # @raise [PuppetForge::V3::Release::ChecksumMismatch] The
       #   cached module release checksum doesn't match the cached checksum.
       #
       # @return [void]
-      def verify_from_md5_file(md5_of_tarball)
-        md5_from_file = File.read(@md5_file_path).strip
-        if md5_of_tarball != md5_from_file
-          logger.error "MD5 of #{@tarball_cache_path} (#{md5_of_tarball}) does not match checksum #{md5_from_file} in #{@md5_file_path}. Removing both files."
-          cleanup_cached_tarball_path
-          cleanup_md5_file_path
+      def verify_from_file(tarball_checksum, checksum_file_path)
+        checksum_from_file = File.read(checksum_file_path).strip
+        if tarball_checksum != checksum_from_file
+          logger.error "Checksum of #{@tarball_cache_path} (#{tarball_checksum}) does not match checksum #{checksum_from_file} in #{checksum_file_path}. Removing both files."
```

Comment:
Do you think it's worth passing in the checksum type here to maintain putting `MD5` vs `SHA256` in the error message?

Reply:
I don't think so, since a) I can't think of a use case that would involve scripting against checksum mismatches and b) the type of the checksum should still be inferable from the path (eg. `puppetlabs-concat-6.0.0.md5` or however that path works out). So I'd say the risk of breaking folks is very low and there's no loss in data for the end user.

But I'd be curious what the community thinks. @binford2k do you know if folks would be upset if we changed the checksum mismatch error message?

Reply:
I've not ever seen anyone write tooling that depends on this. And I don't see a way to do it without changing the message, so I see no blocker here.